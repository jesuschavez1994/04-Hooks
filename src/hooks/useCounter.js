import { useState } from "react";


export const useCounter = ( inicialState = 10 ) => {

    const [state, setstate] = useState(inicialState);

    const increment = ( factor = 2) => {
        setstate( state + factor );
    }

    const decrement = ( factor = 2) => {
        setstate( state - factor );
    }

    const reset = () => {
        setstate( inicialState )
    }

    return {
        state,
        increment,
        decrement,
        reset
    }

}
