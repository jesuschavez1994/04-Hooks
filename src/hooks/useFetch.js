import { useEffect, useRef, useState } from "react"


export const useFetch = ( uri ) => {

    const isMounted = useRef(true);
    const [state, setstate] = useState({data: null, loadin: true, error: null });

    useEffect(() => {
       return () => {
        isMounted.current = false;
       }
    }, [])

   useEffect(() => {
        setstate({data: null, loadin: true, error: null})
        fetch( uri ).then( resp =>  resp.json() ).then( data => {
            if(isMounted.current){
                setstate({
                    loadin: false,
                    error: null,
                    data
                });
            }
        });
   }, [uri])

   return state;
}
