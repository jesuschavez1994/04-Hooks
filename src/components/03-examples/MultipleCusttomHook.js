import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'
import './multiple.css'

export const MultipleCusttomHook = () => {

    const {state: counter, increment} = useCounter( 1 );
    const { loadin, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`) ;
    const {author, quote} = !!data && data[0];



    return (
        <>
            <h1> Multiple Cusstom Hooks</h1>
            <hr />

            {
                loadin ?
                (
                    <div className="alert alert-info text-center">
                        Loading...
                    </div>

                ) 
                
                :

                (
                    <blockquote className="blockquote text-right">
                        <p className="mb-0">{ quote }</p>
                        <footer className="blockquote-footer">{ author }</footer>
                    </blockquote>

                )
            }

            <button className="btn btn-primary mb-5" onClick={ increment }>
                Siguiente
            </button>



        </>
    )
}
