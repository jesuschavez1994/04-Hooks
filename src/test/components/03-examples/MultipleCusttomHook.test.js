
import { shallow } from 'enzyme';
import { MultipleCusttomHook } from '../../../components/03-examples/MultipleCusttomHook';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
import '@testing-library/jest-dom';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultipleCusttomHook />', () => {

    beforeEach( () =>  {
 
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        });
 
    });

    test('Debe de mostrar el componente correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loadin: true,
            error: null
        });

        const wrapper = shallow(<MultipleCusttomHook />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar la informaciÃ³n ', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'Jesus',
                quote: 'Pasiencia libre'
            }],
            loadin: false,
            error: null
        });

        const wrapper = shallow(<MultipleCusttomHook />);
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Pasiencia libre');
        expect(wrapper.find('footer').text().trim()).toBe('Jesus');
    })
    
    
})
