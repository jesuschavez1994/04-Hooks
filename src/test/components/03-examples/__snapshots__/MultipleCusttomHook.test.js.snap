// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCusttomHook /> Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h1>
     Multiple Cusstom Hooks
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary mb-5"
    onClick={[Function]}
  >
    Siguiente
  </button>
</Fragment>
`;

exports[`Pruebas en <MultipleCusttomHook /> should  1`] = `
<Fragment>
  <h1>
     Multiple Cusstom Hooks
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary mb-5"
    onClick={[Function]}
  >
    Siguiente
  </button>
</Fragment>
`;
